host_checker.go

```diff
@@ -104,23 +104,23 @@ func (h *HostUptimeChecker) HostReporter() {
 		case failedHost := <-h.errorChan:
 
 			cachedHostCount, found := h.sampleCache.Get(failedHost.ID)
+
 			if !found {
-				go h.sampleCache.Set(failedHost.ID, 1, cache.DefaultExpiration)
+				cachedHostCount = 0
+			}
 
-			} else {
-				newVal := cachedHostCount.(int)
-				newVal++
-				go h.sampleCache.Set(failedHost.ID, newVal, cache.DefaultExpiration)
-
-				if newVal > h.sampleTriggerLimit {
-					log.Debug("[HOST CHECKER] [HOST WARNING]: ", failedHost.CheckURL)
-					// Reset the count
-					go h.sampleCache.Set(failedHost.ID, 1, cache.DefaultExpiration)
-					// track it
-					h.unHealthyList[failedHost.ID] = true
-					// Call the custom callback hook
-					go h.failureCallback(failedHost)
-				}
+			newVal := cachedHostCount.(int)
+			newVal++
+			go h.sampleCache.Set(failedHost.ID, newVal, cache.DefaultExpiration)
+
+			if newVal >= h.sampleTriggerLimit {
+				log.Debug("[HOST CHECKER] [HOST WARNING]: ", failedHost.CheckURL)
+				// Reset the count
+				go h.sampleCache.Set(failedHost.ID, 1, cache.DefaultExpiration)
```

Comment:
What if the `Set` above happens to run after the one here? Do we really need to fire the two goroutines, ever?

(I know this is not your code and it's not modified in this PR, but still)